core/shared/src/main/scala/scalaz/zio/ExitResult.scala

```diff
@@ -9,56 +10,203 @@ package scalaz.zio
-    case Completed(_) => true
-    case _            => false
+  final def toEither: Either[Throwable, A] = self match {
+    case Completed(value)  => Right(value)
+    case Terminated(cause) => Left(cause.toThrowable())
+  }
+
+  final def leftMap[E1](f: E => E1): ExitResult[E1, A] =
+    self match {
+      case e @ Completed(_) => e
+      case Terminated(c)    => Terminated(c.map(f))
+    }
+
+  final def map[A1](f: A => A1): ExitResult[E, A1] =
+    self match {
+      case Completed(v)      => Completed(f(v))
+      case e @ Terminated(_) => e
+    }
+
+  final def bimap[E1, A1](f: E => E1, g: A => A1): ExitResult[E1, A1] =
+    leftMap(f).map(g)
+
+  final def zip[E1 >: E, B](that: ExitResult[E1, B]): ExitResult[E1, (A, B)] = zipWith(that)((_, _), _ ++ _)
+
+  final def zipPar[E1 >: E, B](that: ExitResult[E1, B]): ExitResult[E1, (A, B)] = zipWith(that)((_, _), _ && _)
+
+  final def zipWith[E1 >: E, B, C](that: ExitResult[E1, B])(
+    f: (A, B) => C,
+    g: (Cause[E], Cause[E1]) => Cause[E1]
+  ): ExitResult[E1, C] =
+    (self, that) match {
+      case (Completed(a), Completed(b))   => Completed(f(a, b))
+      case (Terminated(l), Terminated(r)) => Terminated(g(l, r))
+      case (e @ Terminated(_), _)         => e
+      case (_, e @ Terminated(_))         => e
+    }
+
+  final def exceptions: Option[List[Throwable]] = self match {
+    case Completed(_)      => None
+    case Terminated(cause) => Some(cause.exceptions)
   }
 
-  final def map[B](f: A => B): ExitResult[E, B] = self match {
-    case Completed(a) => Completed(f(a))
-    case x            => x.asInstanceOf[ExitResult[E, B]]
+  final def <>[E1, A1 >: A](that: ExitResult[E1, A1]): ExitResult[E1, A1] = self.failure match {
+    case Some(_) => that
+    case None    => self.asInstanceOf[ExitResult[E1, A1]]
   }
 
-  final def mapError[E2, A1 >: A](f: E => ExitResult[E2, A1]): ExitResult[E2, A1] = self match {
-    case ExitResult.Failed(e, _) => f(e)
-    case x                       => x.asInstanceOf[ExitResult[E2, A1]]
+  final def failure: Option[E] = self match {
+    case ExitResult.Terminated(cause) => cause.failure
+    case _                            => None
   }
 
-  final def failed: Boolean = !succeeded
+  final def succeeded: Boolean = self match {
+    case Completed(_) => true
+    case _            => false
+  }
 
-  final def fold[Z](completed: A => Z, failed: (E, List[Throwable]) => Z, interrupted: List[Throwable] => Z): Z =
+  final def fold[Z](completed: A => Z, terminated: Cause[E] => Z): Z =
     self match {
-      case Completed(v)  => completed(v)
-      case Failed(e, ts) => failed(e, ts)
-      case Terminated(e) => interrupted(e)
+      case Completed(v)      => completed(v)
+      case Terminated(cause) => terminated(cause)
+    }
+}
+object ExitResult {
+
+  final case class Completed[A](value: A)         extends ExitResult[Nothing, A]
+  final case class Terminated[E](cause: Cause[E]) extends ExitResult[E, Nothing]
```

Comment:
I'm increasingly convinced we should call this `Halted`:

```scala
final case class Halted[E](cause: Cause[E]) extends ExitResult[E, Nothing]
```

Either a fiber _completes_ or it _halts_. Either it ends up _completed_ or it ends up _halted_. Halting can occur for different _causes_. Etc.